Add 'semgrep scan --x-ls' subcommand and git-based listing in osemgrep

mjambon opened the pull request:
osemgrep now calls `git ls-files` if a git project was detected.

I still don't have unit tests for this but if the CI tests pass, we should probably merge it to minimize conflicts.

test plan: from within a repo such as the semgrep repo:
```
$ semgrep scan --x-ls | tee ls-py.out | grep '^+' > ls-selected-py.out
$ semgrep scan --x-ls --experimental | tee ls-ml.out | grep '^+' > ls-selected-ml.out
$ diff -u ls-selected-py.out ls-selected-ml.out | less
```

The original output of `--x-ls` (files `ls-py.out` and `ls-ml.out`) shows the files that were selected as well as those that are ignored, with the reason why they're being ignored. The files `ls-selected-py.out` and `ls-selected-ml.out` contain only the selected targets and the diff command shows the differences between pysemgrep and osemgrep. They provide different results and I need to investigate further.

Update: osemgrep no longer ignores dotfiles automatically. 3 tests fail apparently because a `.git` folder isn't being ignored. I'm not sure what mechanism causes it to be ignored in pysemgrep when the current project is _not_ a git project. I'll look into this tomorrow.

Commit:
Simplify and rename Rpath -> Realpath, extend Git_wrapper.ls_files

Commit:
Use git to obtained tracked files when applicable.
No idea why 'test_spacegrep_timeout' gives the wrong output.

Commit:
Fix module rename

Commit:
Add a '--x-ls' option to osemgrep for listing selected and skipped targets

Commit:
Add '--x-ls' option to pysemgrep

Commit:
Work in progress: Two paths for filtering targets depending on whether
it's a git project or not.

Commit:
Almost done with path conversions for getting a list of targets from git

Commit:
Fix type errors

Commit:
Don't ignore dotfiles systematically with osemgrep

aryx commented on cli/src/semgrep/commands/scan.py:
```diff
@@ -365,6 +365,18 @@ def scan_options(func: Callable) -> Callable:
+    is_flag=True,
+    default=False,
+    help=(
+        "[INTERNAL] List the selected target files and the skipped target"
```
the help here is useless because osemgrep is now handling semgrep scan --help

aryx replied:
Better to remove it (I've removed all the help=() in this file a month ago).

mjambon replied:
sure, I noticed it was useless after the fact.

aryx commented on cli/src/semgrep/run_scan.py:
```diff
@@ -300,6 +300,16 @@ def run_rules(
+    for path in sorted(targets.kept):
+        print(f"+ {path}")
+    for (path, reason) in target_manager.ignore_log.list_skipped_paths_with_reason():
+        print(f"- [{reason}] {path}")
```
maybe better to put the reason after the path

mjambon replied:
I did that because I find the output a little easier to read this way, paths being of variable length. It looks like this:
```
- [semgrepignore_patterns_match] tests/tainting_rules/java/try_return.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/await.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/await.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/call_chain.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/call_chain.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/eslint_obj_inj.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/eslint_obj_inj.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/everything_source.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/everything_source.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/everything_source1.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/everything_source1.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/metavar_eq_conditional.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/metavar_eq_conditional.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/metavar_eq_simple.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/metavar_eq_simple.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/nested_function.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/nested_function.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/sanitized_by_side_effect.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/sanitized_by_side_effect.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/simpl_nodejs_eval.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/simpl_nodejs_eval.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/single_variable_source.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/single_variable_source.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/js/throw.js
- [semgrepignore_patterns_match] tests/tainting_rules/js/throw.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/php/break.php
- [semgrepignore_patterns_match] tests/tainting_rules/php/break.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/php/echo.php
- [semgrepignore_patterns_match] tests/tainting_rules/php/echo.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/php/lval_var_sink.php
- [semgrepignore_patterns_match] tests/tainting_rules/php/lval_var_sink.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/php/no_duplicate_submatches.php
- [semgrepignore_patterns_match] tests/tainting_rules/php/no_duplicate_submatches.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/php/switch.php
- [semgrepignore_patterns_match] tests/tainting_rules/php/switch.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/php/ternary.php
- [semgrepignore_patterns_match] tests/tainting_rules/php/ternary.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/break.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/break.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/fake_tokens.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/fake_tokens.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/raise_from.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/raise_from.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/raise_no_args.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/raise_no_args.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/simpl_django_redirect.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/simpl_django_redirect.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/sink_param.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/sink_param.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/sink_return.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/sink_return.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/source_param.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/source_param.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/tainting_top.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/tainting_top.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/try_finally.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/try_finally.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/try_finally1.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/try_finally1.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/python/try_raise.py
- [semgrepignore_patterns_match] tests/tainting_rules/python/try_raise.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/ruby/switch.rb
- [semgrepignore_patterns_match] tests/tainting_rules/ruby/switch.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/scala/new_object_sink.scala
- [semgrepignore_patterns_match] tests/tainting_rules/scala/new_object_sink.yaml
- [semgrepignore_patterns_match] tests/tainting_rules/ts/source_exp.ts
- [semgrepignore_patterns_match] tests/tainting_rules/ts/source_exp.yaml
- [semgrepignore_patterns_match] tests/typing/BasicParam.java
- [semgrepignore_patterns_match] tests/typing/ClassFields.java
- [semgrepignore_patterns_match] tests/typing/EqVarCmp.java
- [semgrepignore_patterns_match] tests/typing/FuncParam.go
- [semgrepignore_patterns_match] tests/typing/PropVarDef.go
- [semgrepignore_patterns_match] tests/typing/StaticVarDef.go
- [semgrepignore_patterns_match] tests/typing/VarDef.java
- [semgrepignore_patterns_match] tmp/foo
- [semgrepignore_patterns_match] tmp/foo.hack
- [semgrepignore_patterns_match] tmp/foo.php
- [semgrepignore_patterns_match] tmp/foo1.js
- [semgrepignore_patterns_match] tmp/foo2.js
- [semgrepignore_patterns_match] tmp/foo3.bar
- [semgrepignore_patterns_match] tmp/hello1.bash
- [semgrepignore_patterns_match] tmp/hello2
- [semgrepignore_patterns_match] tmp/hello3
- [semgrepignore_patterns_match] tmp/hello4
- [semgrepignore_patterns_match] tmp/hello5
- [semgrepignore_patterns_match] tmp/hello6
```

mjambon replied:
I don't love this output format overall but we can change it anytime.

aryx commented on cli/src/semgrep/target_manager.py:
```diff
@@ -176,6 +176,20 @@ def unsupported_lang_paths(self) -> FrozenSet[Path]:
+    def list_skipped_paths_with_reason(self) -> List[Tuple[Path, str]]:
+        res: List[Tuple[Path, str]] = []
+        # The strings used to describe the reason are those defined
+        # in semgrep_output_v1.atd for type 'skip_reason':
```
then would be better to use the atd type there, and just call .to_json() on it later,
so the return type of this function would be List[Tuple[Path, out.SkipReason]]

aryx commented on cli/src/semgrep/run_scan.py:
```diff
@@ -300,6 +300,16 @@ def run_rules(
 
+# This is used for testing and comparing with osemgrep.
+def list_targets_and_exit(target_manager: TargetManager, product: out.Product) -> None:
+    targets = target_manager.get_files_for_language(None, product)
```
We pass None here, but the type says Union[Language, Ecosystem], so how mypy allows this?
should be Optional[Union[Language,Ecosystem]] no?

mjambon replied:
Mypy detects this kind of type error in other contexts. I don't know why it's not detecting the error here. After I changed the signature to `Union[None, Language, Ecosystem]`, it found an error in my code (which was working anyway because of some other type sloppiness?).

aryx commented on cli/src/semgrep/target_manager.py:
```diff
@@ -459,7 +473,7 @@ def files_from_git_ls(self) -> FrozenSet[Path]:
                 "git",
                 "ls-files",
-                "--other",
+                "--others",
```
intended change?
What is the difference?
maybe add a comment then

aryx commented on cli/tests/e2e/snapshots/test_exclude_include/test_exclude_include/exclude0/err.out:
```diff
@@ -3,16 +3,20 @@
 └─────────────┘
-  Scanning 5 files tracked by git with 4 Code rules:
-  Scanning 2 files.
+  Scanning 3 files tracked by git with 4 Code rules:
```
Why the changed snapshot?

aryx commented on libs/git_wrapper/Git_wrapper.mli:
```diff
@@ -17,10 +17,25 @@ type status = {
+   Please extend this interface as needed.
+*)
+type ls_files_kind =
+  | Cached (* --cached, the default *)
```
maybe a small explanation of what those flags mean? With an example?

aryx replied:
(* From git ls-files man page: show all files cached in Git’s index, i.e. all tracked files *)

aryx replied:
For Others (* "Show other (i.e. untracked) files in the output", that is mostly the complement of Cached but still exclusing .git/ *)

aryx commented on libs/git_wrapper/Git_wrapper.mli:
```diff
@@ -17,10 +17,25 @@ type status = {
    This returns a list of paths relative to cwd.
 *)
-val files_from_git_ls : cwd:Fpath.t -> Fpath.t list
+val ls_files :
```
Why do we pass both a cwd and root_paths?

aryx commented on libs/paths/Ppath.mli:
```diff
@@ -51,6 +51,15 @@ val from_segments : string list -> (t, string) result
 (* Convert back to a system path. *)
 val to_fpath : root:Fpath.t -> t -> Fpath.t
 
+(* Convert from a system path to a normalized project path.
```
Why adding those functions? I've changed Ppath.mli to be minimalistic and to not offer many way
to build ppath on purpose. A ppath should be part of a project, hence the in_project function builder above.
For your of_fpath, what will be the resulting project path?

aryx replied:
We should not support those functions because then it's very tempting for people
to make errors and try to get a ppath from an fpath.
To build a ppath, we should clearly pass a root and then an fpath, not just an fpath!

aryx commented on libs/paths/Project.ml:
```diff
@@ -1,8 +1,17 @@
+
+   A git project must work with the 'git' commands.
+
+   A gitignore project is a fake git project. It doesn't have a valid
```
seems weird to have a .gitignore but no .git
Which project have one but not the other?

aryx commented on libs/paths/Realpath.ml:
```diff
@@ -38,27 +38,14 @@
 (* Types *)
 (*****************************************************************************)
 
-(* TODO? use of Fpath.t instead? or too confusing? *)
```
please do not rename to Realpath.ml. I used Rpath on purpose to match Fpath, Ppath.

aryx commented on libs/paths/Realpath.mli:
```diff
@@ -20,16 +20,13 @@
    findings relative to the root of their project (and not real paths like
    /home/pad/my/long/project/foo/bar).
-
-   The name of the module imitates Fpath.ml, and Ppath.ml, but use Rpath.ml
```
See, it was done on purpose!

mjambon replied:
I have no doubt it was done on purpose but it is and should remain an uncommon call due to its cost, and uncommon things should have longer, more descriptive names.

aryx commented on src/osemgrep/cli_ci/Ci_subcommand.ml:
```diff
@@ -703,7 +703,7 @@ let run_conf (ci_conf : Ci_CLI.conf) : Exit_code.t =
     let profiler = Profiler.make () in
     let targets_and_ignored =
-      Find_targets.get_targets conf.targeting_conf conf.target_roots
+      Find_targets.get_target_fpaths conf.targeting_conf conf.target_roots
```
Why the _fpaths suffix? We are using more consistently fpaths now in semgrep, so we should assume
target filenames have Fpath.t type.
Or is it because we also define some Input_to_core.target and xtarget type?

mjambon replied:
No, it's because the default `get_targets` function now returns fppaths.

mjambon replied:
I thought fppaths could be useful for troubleshooting or for further filtering. Maybe it's confusing?

aryx commented on src/osemgrep/cli_scan/Ls_subcommand.ml:
```diff
@@ -0,0 +1,26 @@
+open Common
+module Out = Semgrep_output_v1_j
+
+let string_of_json_string json =
```
You can reuse JSON.remove_enclosing_quotes_of_jstring
which raise NotAJString

aryx commented on src/osemgrep/cli_scan/Ls_subcommand.mli:
```diff
@@ -0,0 +1,19 @@
+   currently implemented as a '--ls' flag of 'semgrep scan' for simplicity
+   of implementation.
+   If we wanted to make it official and permanent, we should probably
+   turn it into a proper 'semgrep ls' subcommand.
```
or add it to  'semgrep show targets'

aryx commented on src/osemgrep/cli_scan/Ls_subcommand.mli:
```diff
@@ -0,0 +1,19 @@
+   or language-specific filtering.
+
+   This is an internal/experimental option for troubleshooting
+   currently implemented as a '--ls' flag of 'semgrep scan' for simplicity
```
--ls or -x-ls?

mjambon replied:
`--x-ls` (I originally wanted to call it just `--ls`)

aryx commented on src/targeting/Find_targets.ml:
```diff
@@ -109,6 +105,60 @@ let get_reason_for_exclusion (sel_events : Gitignore.selection_event list) :
 (* Finding *)
 (*************************************************************************)
 
+type filter_result =
```
nice type

aryx reviewed (changes requested):
Looks mostly good, but I'd rather get back Rpath.ml and the smaller Ppath.mli

Commit:
Remove unused help message on the Python side

Commit:
Use lib function

Commit:
Fix comment

Commit:
Add comments

Commit:
Remove tricky function from Ppath interface

Commit:
Comment

Commit:
Remove dead code

Commit:
Revert the module rename back from Realpath to Rpath to avoid having
to fix the incompatibility with semgrep-pro.

Commit:
Fix type error that was undetected by mypy (why?)

Commit:
Merge remote-tracking branch 'origin/develop' into martin/osemgrep-git-ls

Commit:
Fix rename

Commit:
Ignore .git

Commit:
Update test snapshot
I'm not sure why they changed but they look like the output of the other
tests in the same series.

aryx commented on src/targeting/Semgrepignore.ml:
```diff
@@ -59,6 +59,9 @@ type exclusion_mechanism = Gitignore_and_semgrepignore | Only_semgrepignore
 *)
 let builtin_semgrepignore_for_semgrep_scan =
   {|
+# Git administrative folder or file
```
could be good to add why we put this folder given git --others is skipping this directory already
and why pysemgrep does not need this .git in his own .semgrep_ignore template file.
